src/app/icd/ICDCheckInSender.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDCheckInSender.h"
+
+#include <system/SystemPacketBuffer.h>
+
+#include <protocols/secure_channel/CheckinMessage.h>
+
+#include <lib/dnssd/Resolver.h>
+
+namespace chip {
+namespace app {
+
+using namespace Protocols::SecureChannel;
+
+ICDCheckInSender::ICDCheckInSender(Messaging::ExchangeManager * exchangeManager)
+{
+    VerifyOrDie(exchangeManager != nullptr);
+    mExchangeManager = exchangeManager;
+    mAddressLookupHandle.SetListener(this);
+}
+
+void ICDCheckInSender::OnNodeAddressResolved(const PeerId & peerId, const AddressResolve::ResolveResult & result)
+{
+    mResolveInProgress = false;
+
+    VerifyOrReturn(CHIP_NO_ERROR != SendCheckInMsg(mCurrentEntry, result.address),
+                   ChipLogError(AppServer, "Failed to send the ICD Check-In message"));
+}
+
+void ICDCheckInSender::OnNodeAddressResolutionFailed(const PeerId & peerId, CHIP_ERROR reason)
+{
+    mResolveInProgress = false;
+    ChipLogProgress(AppServer, "Node Address resolution failed for ICD Check-In with Node ID " ChipLogFormatX64,
+                    ChipLogValueX64(peerId.GetNodeId()));
+}
+
+CHIP_ERROR ICDCheckInSender::SendCheckInMsg(ICDMonitoringEntry & entry, const Transport::PeerAddress & addr)
+{
+
+    System::PacketBufferHandle buffer = MessagePacketBuffer::New(CheckinMessage::sMinPayloadSize);
+
+    VerifyOrReturnError(!buffer.IsNull(), CHIP_ERROR_NO_MEMORY);
+    MutableByteSpan output{ buffer->Start(), buffer->DataLength() };
+
+    // TODO retrieve Check-in counter
+    CounterType counter = 0;
+
+    ReturnErrorOnFailure(CheckinMessage::GenerateCheckinMessagePayload(entry.key, counter, ByteSpan(), output));
+
+    VerifyOrReturnError(mExchangeManager->GetSessionManager() != nullptr, CHIP_ERROR_INTERNAL);
+
+    // Using default MRP since we are not doing MRP in this context
+    Messaging::ExchangeContext * exchangeContext = mExchangeManager->NewContext(
+        mExchangeManager->GetSessionManager()->CreateUnauthenticatedSession(addr, GetDefaultMRPConfig()).Value(), nullptr);
```

Comment:
This crashes if CreateUnauthenticatedSession returns a NullOptional.  Is that intended?

Reply:
fixed